WalletWasabi/Services/WalletService.cs

```diff
@@ -1068,6 +1073,68 @@ public IEnumerable<string> GetNonSpecialLabels()
 				.Distinct();
 		}
 
+		private async Task RefreshHistoryAsync(SemaphoreSlim semaphore, SmartCoin coin, ILookup<Script, SmartCoin> lookupScriptPubKey, ILookup<uint256, SmartCoin> lookupSpenderTransactionId, ILookup<uint256, SmartCoin> lookupTransactionId)
+		{
+			if (semaphore != null)
+			{
+				await semaphore.WaitAsync();
+			}
+			var result = string.Join(", ", GetHistory(coin, new List<SmartCoin>(), lookupScriptPubKey, lookupSpenderTransactionId, lookupTransactionId).Select(x => x.Label).Distinct());
+			coin.SetHistory(result);
```

Comment:
I don't know if we have to display a very large history. I should think more about this but at least in the UI we cannot see all the history without scrolling right. If we don't show all the history then we don't need to calculate all the history and can stop after some level.

Reply:
And then you accidentally mix you silk road coins with your mtgox coins, since the wallet didn't show crossed history?:) We'll have to show clusters to the user later, but that's a large work and a lot of figuring things out.

Reply:
Yes, you're right but my point was not about saving some CPU calculation. My point is to improve in some way the history. I think that displaying the cluster would be the best by sure but there could be something to do that don't require a big effort like allowing to edit/override a label or shorting the text in some way.

Reply:
I see. I still don't think that's a good idea. The history is hard to understand it's an issue, but if it's not accurate, that's a larger issue.

Reply:
My opinion is for now it is good like that.